### PR TITLE
ENH: create FlightDataImporter class

### PR DESCRIPTION
**Pull request by Gui-FernandesBR**
## Pull request type

- [x] Code changes (bugfix, features)

## Checklist

- [x] Tests for the changes have been added (if needed)
- [x] Docs have been reviewed and added / updated
- [x] Lint (`black rocketpy/ tests/`) has passed locally 
- [x] All tests (`pytest --runslow`) have passed locally

## Current behavior
See issue #284 . 

## New behavior
I created this `DataFrameToFlight` class that I'd like to discuss before going to develop. I'm still struggling to find the best architecture for this, but at least it is a first step.

## Breaking change

- [x] No

### CODE REVIEW

**Comment by Gui-FernandesBR** [comment]
To see an example of the new class being used, see this branch: https://github.com/RocketPy-Team/RocketPy/tree/enh/add-juno3-example

File: rocketpy/simulation/df2flight.py
Comment (giovaniceotto) [inline rocketpy/simulation/df2flight.py:120]: Pandas is quite a heavy library (> 10 MB while RocketPy is < 300 kB). So, adding it as import (which will significantly slow down importing this file) just to define the type of a data class attribute seems a bit exaggerated to me.

In fact, rumor says that even if you don't import pandas, this line still runs fine since python completely ignores type annotations, as long as what is written is correct syntax wise (see https://death.andgravity.com/dataclasses).

So, I suggest two options:

1. Don't import pandas and keep `df: pd.DataFrame`. Works for python, may cause problems with linters.
2. Don't import pandas and change `df: pd.DataFrame` to `'df: pd.DataFrame'`. Works for python, has less chance of causing problems with linters.

```suggestion
    df: 'pd.DataFrame'
```

File: rocketpy/simulation/df2flight.py
Comment (giovaniceotto) [inline rocketpy/simulation/df2flight.py:14]:
```suggestion
```

File: rocketpy/simulation/df2flight.py
Comment (giovaniceotto) [inline rocketpy/simulation/df2flight.py:17]: If this decorator only works inside the `DataFrameToFlight` class, wouldn't it be better to move it there?

See https://www.geeksforgeeks.org/creating-decorator-inside-a-class-in-python/

Also, consider other names such as `df_2_function_property` or similar. I feel we are missing the Function class in the name, since it is core part of this decorator.

File: rocketpy/simulation/df2flight.py
Comment (giovaniceotto) [inline rocketpy/simulation/df2flight.py:48]:
```suggestion
        """A decorator class that wraps a method to create a property from a data
        frame, as a Function instance.
        This is made via class because it's the only way to use the __set_name__
        method. This method is used to get the name of the property in the class.
```

File: rocketpy/simulation/df2flight.py
Comment (giovaniceotto) [inline rocketpy/simulation/df2flight.py:84]:
```suggestion
    property of a Flight object will be created using values from the data frame.
```

File: rocketpy/simulation/df2flight.py
Comment (giovaniceotto) [inline rocketpy/simulation/df2flight.py:109]: Does the default option currently work? What happens if you leave it as default?

**Review by giovaniceotto — CHANGES_REQUESTED** [review]
Great addition @Gui-FernandesBR! Can't wait to start using this more!

In fact, I hope we get a similar version that works directly with CSV files and other file formats common for rocket data loggers. I'll reach out to the community to see what kind of files we should support in a more long term scope.

I was pretty close to approving this PR as is, however, I believe I can help improve the implementation just a bit with two comments:

1. Prefer a more explicit file name: `data_frame_to_flight.py`
2. Avoid importing pandas, since it is a big module and will slow down importing this module. When you analyze the code, you really don't need pandas at all, even though you are working with pandas directly (Python is weird...)

File: rocketpy/simulation/df2flight.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/df2flight.py]: Good catch, I fixed this based on your comment.

File: rocketpy/simulation/df2flight.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/df2flight.py]: Yes, it would be a better idea.

But the decorator no longer exists.

**Comment by Gui-FernandesBR** [comment]
For future reference... 

Something that is getting into my nerves is that sometimes the flight data is separated into different files. This may cause differences in the time steps and others.

I guess that an easy solution would be converting the current private method `__create_attributes` to a public one, so the user could call it multiple times if needed. I think that could be a brilliant solution, but I'm not able to test it immediately so saving this for later.

Example:

image

File: rocketpy/simulation/flight_data_importer.py
Comment (MateusStano) [inline rocketpy/simulation/flight_data_importer.py:171]: The way you are iterating through the `columns_map` seem to inverted to the Juno3 example and to the `FLIGHT_LABEL_MAP`. The way its done here, the key is the name of the column in the file and the value is the name of the Flight attribute

Seems like it was intended to be the other way around (which I think is better also)

File: rocketpy/simulation/flight_data_importer.py
Comment (MateusStano) [inline rocketpy/simulation/flight_data_importer.py:47]: I wonder this could be a list of filepaths and, as long as they have headers and a "time" column, it should be easy to make it work

**Review by MateusStano — COMMENTED** [review]
So overall it seems the user needs two things to use this class:

1 - A file with data in a valid format
2 - A dictionary mapping his data to flight values

I would say that makes a lot of sense, we just need a lot of documentation to support these requirements

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py:171]: You are right, I inverted that without noticing. 
The point is that the idea of "mapping" the columns is to get the corresponding Flight attribute for each column name of the dataset. I understand if may not be so beautiful, but I think if makes more sense to have something like `{"column_name": "Flight attribute name", ...}` than the opposite.

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: That's right.

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: Good suggestions, done!

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: I still need to update the juno3 example

File: rocketpy/simulation/df2flight.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/df2flight.py]: pandas is no longer necessary

**Comment by Gui-FernandesBR** [comment]
@RocketPy-Team/code-owners I believe we are ready for a review now.
I'm still working on supplying an extended example on how to use the class and create good comparisons, but the `FlighDataImporter` itself is ready, I believe so.

**Comment by Gui-FernandesBR** [comment]
Please take a look at #513 to see an example of this class being used.

File: rocketpy/simulation/flight_data_importer.py
Comment (MateusStano) [inline rocketpy/simulation/flight_data_importer.py:96]: `delimiter` is a more common name for this

File: rocketpy/simulation/flight_data_importer.py
Comment (MateusStano) [inline rocketpy/simulation/flight_data_importer.py:143]: Do we follow any conventions when naming methods meant to be private? Sometimes we do it with one underscore, sometimes we do it with two.

From what I've researched it is more common to use only one underscore for the use case of making a method "private" or "protected"

File: rocketpy/simulation/flight_data_importer.py
Comment (MateusStano) [inline rocketpy/simulation/flight_data_importer.py:81]: This can also be a string, not just list

**Review by MateusStano — APPROVED** [review]
I think most comments I made are regarding documentation, so feel free to merge if you want to fix those in another PR.

Everything else seems good

**Comment by Gui-FernandesBR** [comment]
> I think most comments I made are regarding documentation, so I am approving this one.
> 
> But unless you are planning to add proper documentation soon, I do not feel comfortable merging this to develop just yet

What do you mean by "proper documentation"?

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py:143]: Writing "Class._method" and "Class.__method" are not the same thing.

Private methods in python actually doesn't exist.
Using double underscore will make python "to mangle" the attribute, which is basically adding a "__ClassName__" at the beginning of the attribute's name, which happens to be a uncomfortable way of "privatizing" attributes and methods within a class.

Using a single underscore is the informal way python developers found to define "private methods" without mangling its name.

I left these methods with double underscores because I don't think users (or developers) should invoke these methods out of this class. I mean, I wanted them private, not "weakly private".

File: rocketpy/simulation/flight_data_importer.py
Comment (MateusStano) [inline rocketpy/simulation/flight_data_importer.py:80]: Seems, that this can be a path to a folder, and thin will read all the .txt and .csv files in there, right?

How does this behave with the `columns_map` argument?

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: The juno3 notebook was updated in another PR

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: nice catch. Solved  here: [4a473d6](https://github.com/RocketPy-Team/RocketPy/pull/474/commits/4a473d6a7a5bb48057917b308e37b39244e8e889)

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: I would say "outdated" instead of "wrong".

Solved here: [4a473d6](https://github.com/RocketPy-Team/RocketPy/pull/474/commits/4a473d6a7a5bb48057917b308e37b39244e8e889)

**Comment by MateusStano** [comment]
> What do you mean by "proper documentation"?

My bad, I meant to delete that srry

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: My initial answer was gonna be:
- Yes, because for a single rocket flight I'm expecting to have different FlightDataImport and and want to ensure the user is naming the objects correctly.

But tbh, there's no hard reason to have this as a mandatory argument. The code should still run without it. 
So I changed it here: [c7ec3ce](https://github.com/RocketPy-Team/RocketPy/pull/474/commits/c7ec3ceb55ce741ff1ef0bba183577b56c4928b2)

Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: Thank you for noticing it.

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: I gave a better description with this commit: [4a473d6](https://github.com/RocketPy-Team/RocketPy/pull/474/commits/4a473d6a7a5bb48057917b308e37b39244e8e889)

Also, the #513 is going to add a good example that will help understanding how to use this class.

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: Good catch. The numpy module also uses delimiter, so I've changed in my latest commit.

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: Solved.

File: rocketpy/simulation/flight_data_importer.py
Comment (Gui-FernandesBR) [reply rocketpy/simulation/flight_data_importer.py]: Yes it can indeed read all the files within a folder, at once.
In this case the same columns_map will be used for all the files.
Again this will become clearer after we merge example notebooks, but basically it is way easier to define the same mapping instead of one particular map for each file.

I ensured the code is working even if you mess up the map... So no worries.

**Pull request merged**
